### PR TITLE
Initial build infrastructure and example for WASM FFI

### PR DESCRIPTION
**Pull request by shadaj**
Adjust the WASM build config to build the FFI package instead of examples, and bring in a simple example of using the FFI from JavaScript.

### CODE REVIEW

File: ffi/capi/src/custom_writeable.rs [hunk 1]
```diff
@@ -116,3 +116,56 @@ pub unsafe extern "C" fn icu4x_simple_writeable(buf: *mut u8, buf_size: usize) -
         grow,
     }
 }
+
+
+#[no_mangle]
+pub unsafe extern "C" fn icu4x_buffer_writeable(cap: usize) -> *mut ICU4XWriteable {
```
Comment (Manishearth) [inline ffi/capi/src/custom_writeable.rs:122]: issue: Please add doc comments to these, highlighting any safety constraints and relations with other functions ("use this function to call destructors", etc)

File: ffi/capi/src/custom_writeable.rs [hunk 2]
```diff
@@ -116,3 +116,56 @@ pub unsafe extern "C" fn icu4x_simple_writeable(buf: *mut u8, buf_size: usize) -
...
+        flush: flush,
+        grow: grow
+    };
+
+    std::mem::forget(vec);
+    Box::into_raw(Box::new(ret))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn icu4x_buffer_writeable_borrow(this: *mut ICU4XWriteable) -> *mut u8 {
```
Comment (Manishearth) [inline ffi/capi/src/custom_writeable.rs:150]: suggestion: _get_bytes maybe?

File: ffi/capi/src/custom_writeable.rs [hunk 3]
```diff
@@ -116,3 +116,56 @@ pub unsafe extern "C" fn icu4x_simple_writeable(buf: *mut u8, buf_size: usize) -
...
+#[no_mangle]
+pub unsafe extern "C" fn icu4x_buffer_writeable_len(this: *mut ICU4XWriteable) -> usize {
+    let this = Box::from_raw(this);
+    let ret = this.len;
+    std::mem::forget(this);
+    ret
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn icu4x_buffer_writeable_free(this: *mut ICU4XWriteable) {
```
Comment (Manishearth) [inline ffi/capi/src/custom_writeable.rs:166]: suggestion: _destroy to be more in line with the other _destroy methods?

File: ffi/capi/src/fixed_decimal.rs [hunk 4]
```diff
@@ -35,6 +38,12 @@ pub extern "C" fn icu4x_fixed_decimal_multiply_pow10(
 pub extern "C" fn icu4x_fixed_decimal_negate(fd: &mut ICU4XFixedDecimal) {
     fd.negate()
 }
+
+#[no_mangle]
+pub extern "C" fn icu4x_fixed_decimal_write_to(fd: &ICU4XFixedDecimal, to: &mut ICU4XWriteable) {
```
Comment (Manishearth) [inline ffi/capi/src/fixed_decimal.rs:43]: issue: doc comments please

**Review by Manishearth — COMMENTED** [review]
At a quick glance the code looks good, but before landing we should have a lot more comments and doc comments where possible.

File: ffi/capi/src/custom_writeable.rs [hunk 5]
```diff
@@ -119,7 +119,11 @@ pub unsafe extern "C" fn icu4x_simple_writeable(buf: *mut u8, buf_size: usize) -
 
 
 #[no_mangle]
-pub unsafe extern "C" fn icu4x_buffer_writeable(cap: usize) -> *mut ICU4XWriteable {
+/// Create an [`ICU4XWriteable`] that can write to a dynamically allocated buffer managed by Rust.
+///
+/// # Safety
+/// - Use [`icu4x_buffer_writeable_destroy()`] to free the writable and its underlying buffer.
```
Comment (Manishearth) [inline ffi/capi/src/custom_writeable.rs:124]: nit: This isn't necessary for safety, can just be a recommendation.

File: ffi/capi/src/custom_writeable.rs [hunk 6]
```diff
@@ -147,14 +151,25 @@ pub unsafe extern "C" fn icu4x_buffer_writeable(cap: usize) -> *mut ICU4XWriteab
...
 
 #[no_mangle]
-pub unsafe extern "C" fn icu4x_buffer_writeable_borrow(this: *mut ICU4XWriteable) -> *mut u8 {
+/// Grabs a pointer to the underlying buffer of a writable.
+///
+/// # Safety
+/// - The returned pointer is valid until the passed writable is destroyed.
+/// `this` must be a pointer to a valid [`ICU4XWriteable`] constructed by
+/// [`icu4x_buffer_writeable_create()`].
+pub unsafe extern "C" fn icu4x_buffer_writeable_get_bytes(this: *mut ICU4XWriteable) -> *mut u8 {
```
Comment (Manishearth) [inline ffi/capi/src/custom_writeable.rs:159]: suggestion: `this` can just be `&ICU4XWriteable` here, so no need for the function itself to be unsafe, though of course the returned buffer still has safety requirements.

File: ffi/capi/src/custom_writeable.rs [hunk 7]
```diff
@@ -116,3 +116,75 @@ pub unsafe extern "C" fn icu4x_simple_writeable(buf: *mut u8, buf_size: usize) -
...
+    ret
+}
+
+#[no_mangle]
+/// Gets the length in bytes of the content written to the writable.
+///
+/// # Safety
+/// - `this` must be a pointer to a valid [`ICU4XWriteable`] constructed by
+/// [`icu4x_buffer_writeable_create()`].
+pub unsafe extern "C" fn icu4x_buffer_writeable_len(this: *mut ICU4XWriteable) -> usize {
```
Comment (Manishearth) [inline ffi/capi/src/custom_writeable.rs:172]: suggestion: ditto about using `&ICU4XCustomWriteable` and no `unsafe`

File: ffi/capi/src/fixed_decimal.rs [hunk 8]
```diff
@@ -35,6 +38,13 @@ pub extern "C" fn icu4x_fixed_decimal_multiply_pow10(
 pub extern "C" fn icu4x_fixed_decimal_negate(fd: &mut ICU4XFixedDecimal) {
     fd.negate()
 }
+
+#[no_mangle]
+/// FFI version of [`FixedDecimal::write_to()`]. See its docs for more details.ICU4XFixedDecimal
```
Comment (Manishearth) [inline ffi/capi/src/fixed_decimal.rs:43]:
```suggestion
/// FFI version of [`FixedDecimal::write_to()`]. See its docs for more details.
```

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
r=me, make sure the code passes clippy and rustfmt

**Review by Manishearth — APPROVED** [review]

File: wasm-example/icu4x.mjs [hunk 9]
```diff
@@ -0,0 +1,31 @@
+let icu4x;
```
Comment (sffc) [inline wasm-example/icu4x.mjs:1]: Nit: Please move these example files to a subdirectory somewhere under *ffi* so as not to clutter the top-level namespace.

**Review by sffc — COMMENTED** [review]
LGTM except for one thing

**Pull request merged**